### PR TITLE
Modify dlc/updates and notify when they are moved or deleted

### PR DESCRIPTION
**Pull request by jms-c**
This will allow DLC's to be moved without issues (other than having to re-add dlcs).
This should solve part of #1366.

### CODE REVIEW

**Comment by jms-c** [comment]
Feel free to change DLC's to DLCS (if you prefer, I don't know which is more grammatically correct)

File: Ryujinx/Ui/Windows/DlcWindow.cs [hunk 1]
```diff
@@ -75,6 +75,13 @@ private DlcWindow(Builder builder, VirtualFileSystem virtualFileSystem, string t
 
             foreach (DlcContainer dlcContainer in _dlcContainerList)
             {
+                if (!File.Exists(dlcContainer.Path))
+                {
+                    Console.WriteLine("DLC's have been deleted or moved, deleting " + _dlcJsonPath);
+                    GtkDialog.CreateErrorDialog("Your DLC's have been moved or deleted, please re-add them.");
```
Comment (EmulationFanatic) [inline Ryujinx/Ui/Windows/DlcWindow.cs:81]: "DLC files have been moved or deleted; please re-add your DLC."
Again sidesteps the apostrophe issue. Your call, but this is more correct in terms of English rules. In terms of punctuation, I'm not sure how it is normally used in dialog boxes and whether a period is correct or whether this should all be in title case.

**Comment by jms-c** [comment]
I will commit some changes to fix those issues.

**Comment by jms-c** [comment]
A new commit to change the wording has been added.
(based on EmulationFanatic's feedback with modifications)

**Comment by jms-c** [comment]
Skips remaining dlc if game is running with invalid dlc and logs in console.

**Comment by jms-c** [comment]
It now logs properly instead of Console.Writeline!

**Comment by jms-c** [comment]
This just closes the dlc manager now after the error, reopen to see empty list (prevents partial list).

**Comment by jms-c** [comment]
This new change makes sure the list stay empty after clear allowing re-adding of dlc easily.
It just now needs some testing, in my opinion. (and gui warning on game launch (but I can't implement that atm because it crashes))

**Comment by EmulationFanatic** [comment]
Works well on Windows. Tested the following:

- Scenario 1: Managing DLC for a game whose DLC has been moved or deleted.
As expected, there was a dialog box warning me and then the DLC manager popped up for me to re-add the DLC file.
- Scenario 2: Launching a game whose DLC has been moved or deleted.
As expected, the DLC was discarded, the event noted in the console & log, and the game launched successfully.

Hopefully adding a pop-up on launch in Scenario 2 instances is possible somehow. That way the user is sure to know that their DLC is not working. If not now, then when the new UI hits :)

**Comment by jms-c** [comment]
> 
> 
> Works well on Windows. Tested the following:
> 
>     * Scenario 1: Managing DLC for a game whose DLC has been moved or deleted.
>       As expected, there was a dialog box warning me and then the DLC manager popped up for me to re-add the DLC file.
> 
>     * Scenario 2: Launching a game whose DLC has been moved or deleted.
>       As expected, the DLC was discarded, the event noted in the console & log, and the game launched successfully.
> 
> 
> Hopefully adding a pop-up on launch in Scenario 2 instances is possible somehow. That way the user is sure to know that their DLC is not working. If not now, then when the new UI hits :)

The gui would be great but any attempt crashes the emulator for some reason (atleast for me).

**Comment by jms-c** [comment]
New minor change, doesn't load any dlc if a invalid dlc is detected. This should be more noticeable

**Comment by jms-c** [comment]
Major Update!
Now it no longer destroys dlc.json but simply removes the invalid entries.

**Comment by jms-c** [comment]
Testing requested. (especially on linux)
https://w.ryujinx.org/dl/2054

**Comment by vickyorlo** [comment]
Tested on Linux:

- Game is Project Diva Mega Mix, with several of the song packs already added in the DLC manager.
- Song Pack 1, containing a song called Double Lariat, moved away to a different folder.
- DLC manager when opened warns about missing DLCs. DLC missing from the list. Pressed the cancel button.
- Opened the DLC manager again, warns about missing DLCs again. Pressed the save button.
- Opened the DLC manager again, the DLC is still missing as expected.
- Added the Song Pack 1 DLC again, and then moved it somewhere else.
- Opened the game, the message about a missing DLC is visible in the console log.
- Double Lariat is gone from the song list, but songs from the other DLC packs are still visible.
- Closed the game and opened the DLC manager, warns about missing DLC.
- Added the Song Pack 1 DLC again.
- Double Lariat is visible again.

To summarize, everything works as expected.

**Comment by jms-c** [comment]
If anyone could submit working gui code, it would be helpful as using "_device.UiHandler.DisplayMessageDialog" causes the emulator to freeze.
Therefore, It isn't in the pr.

**Comment by jms-c** [comment]
I added a gui warning. There is however an issue as it doesn't dispose _emulationContext (it hangs if i try).

**Comment by jms-c** [comment]
It will now restart ryujinx if user responds no on ui.

**Comment by jms-c** [comment]
Please test the new gui!
https://w.ryujinx.org/dl/2054
Try to launch the game after moving a dlc file.
Comment your results.

File: Ryujinx/Ui/MainWindow.cs [hunk 2]
```diff
@@ -486,32 +487,51 @@ public void LoadApplication(string path)
...
+                                {
+                                    Logger.Error?.Print(LogClass.Application, "The file which you have specified is unsupported by Ryujinx.");
+                                }
+                                break;
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.Error?.PrintMsg(LogClass.Loader, e.Message);
+                        bool continueGame = GtkDialog.CreateChoiceDialog("An error occurred while trying to load the game!",
```
Comment (EmulationFanatic) [inline Ryujinx/Ui/MainWindow.cs:523]: For the title bar on this window, something more generic in title case like "Game Load Error" would be preferred imho.

File: Ryujinx/Ui/MainWindow.cs [hunk 3]
```diff
@@ -486,32 +487,51 @@ public void LoadApplication(string path)
...
+                           e.Message, "Would you like to continue playing? (pressing no will restart ryujinx)");
```
Comment (EmulationFanatic) [inline Ryujinx/Ui/MainWindow.cs:524]: "Continue loading anyway? Pressing "No" will restart the emulator."

File: Ryujinx/Ui/MainWindow.cs [hunk 4]
```diff
@@ -486,32 +487,51 @@ public void LoadApplication(string path)
...
+                        if (!continueGame)
+                        {
+                            Logger.Warning?.PrintMsg(LogClass.Application, "Stopped loading game due to an exception!");
```
Comment (EmulationFanatic) [inline Ryujinx/Ui/MainWindow.cs:527]: "Game launch was aborted due to an exception"

**Review by EmulationFanatic — COMMENTED** [review]
Works as intended on Windows here. I tested all scenarios I could think of and everything worked as it should.
Added some verbiage suggestions but they are your call to adopt or discard :)

**Comment by jms-c** [comment]
Thanks for the improvements. I have updated the wording (imo, it's clearer)!

File: Ryujinx/Ui/MainWindow.cs (hunk 3 above)
Comment (jms-c) [reply Ryujinx/Ui/MainWindow.cs]: I like that.
The "Continue loading anyway" makes it sound more like "Their is a problem, stop running" instead of "You can continue - I don't mind"

**Comment by vickyorlo** [comment]
Tested again on Linux:

- First five steps the same
- Running the game shows the DLC missing warning. Pressed "No".
- Ryujinx restarts.
- Launched the game again, pressing "yes"
- Double Lariat not present, songs from other song packs present.

Once again, looks good.

**Comment by jms-c** [comment]
I think this PR is ready, but if any last minute testers want to test feel-free,
To test, install dlc using the manager then move the file using a file manager.
It shouldn't crash when editing dlc and launching the game.

Here is a shiny download button:
[Download Ryujinx](https://w.ryujinx.org/dl/2054)

Please comment recomendations and results on this pr.

**Comment by EmulationFanatic** [comment]
As fluid as this functionality is, I have reservations about the entire emulator restarting. I think finding a way to properly stop the loading (or check for DLC prior to launching the game) would be preferred if at all possible.

**Comment by jms-c** [comment]
I would love to stop the emulator but the only way I see is _emulationContext.dispose(); return; but the disposal hanged the emulator on testing.

**Comment by EmulationFanatic** [comment]
> I would love to stop the emulator but the only way I see is _emulationContext.dispose(); return; but the disposal hanged the emulator on personal testing.

What about a pre-load DLC check to prevent launch in the first place?

**Comment by jms-c** [comment]
If I can get the titleId it should be possible.

**Comment by EmulationFanatic** [comment]
> If I can get the titleId it should be possible. (early enough)
> However, it looks like the titleId is only initialized after loading.

Is there a sensible way of changing that? To identify the titleID prior to loading, that is.

**Comment by jms-c** [comment]
It requires parsing nsp/xci headers it seems. which = lots of copied code.

**Comment by mageven** [comment]
The suggested `UiHandler` code didn't work because, for some reason, `LoadApplication()` executes on the GUI thread. This is also part of the reason the GUI completely freezes when you start loading a game. But making it run on a different thread is a disruptive change and better off as its own PR.

Restarting the whole emulator application is a bad idea and should be avoided at all costs (especially after detecting the problem).

Fwiw, I thought your earlier solution to just log a warning and proceed wasn't bad. It remains the simplest solution for now.

**Comment by jms-c** [comment]
I added a gui as running the game causes undefined behavour with some games completely deleting parts of saves due to missing dlc. That is why I offer the option of stopping the game before it gets to a point where that happens. (otherwise people might not be very happy)

**Comment by vickyorlo** [comment]
See my previous report: 

> However, I noticed the score I had on Double Lariat got erased. Perhaps some more visible warning could be useful to prevent "undefined behavior" when a game overwrites a save with missing DLCs.

Various games might handle missing DLCs in different ways, in my opinion a pop-up warning is fairly useful in terms of user experience. The warning in the console instantly gets flooded by other messages.

**Comment by jms-c** [comment]
I think this is ready. Feel free to review whenever.

**Comment by MetrosexualGarbodor** [comment]
Tested on Windows 10: ran the game (Xenoblade 2) to check the DLC and update were installed, closed the emulator, moved the DLC *and* the update to another folder, opened Ryujinx, clicked on "Manage DLC" and it gave me the following error:
unknown (1)
So it would seem it is working as intended. 
My only feedback past that would be considering if there should be (or if it's even worth implementing) a similar functionality for when updates are moved, as stuff can break when the save file and DLC are on, but the updates are off. 
I also tested if it would say anything different if I only added 5 out of 6 DLCs, and it didn't. I realize this falls on the user if they somehow forget about one file, but I believe it's also worth considering if it's worth implementing for the future. 
Thank you for your work.

**Comment by jms-c** [comment]
If you want to test here: 

Here is a shiny download button:
[Download Ryujinx](https://w.ryujinx.org/dl/2054)

Please comment recomendations and results on this pr.

**Comment by jms-c** [comment]
If anyone wants to review, please do.

**Comment by jms-c** [comment]
Do I need to do anything more for this pull request?

**Comment by jms-c** [comment]
Please review.

**Review by AcK77 — CHANGES_REQUESTED** [review]
Why you don’t remove the missing DLC file in the list and then run the game without the missing DLC instead of throw an exception (It’s a bad practice in this situation), restart Ryujinx (I don’t see why you want to restart the emulator), and more, which is not a good solution to fixes the issue.

Nice try for a first contribution! Keep it up!

**Comment by jms-c** [comment]
I found an issue that happend when I previously did start the game with missing DLC and a console warning.
The issue is due to how games handle missing DLC. Starting some games (especially bad with autosaving) may do varying things from ignoring dlc (and using save) or deleting data (from dlc) and more

**Comment by jms-c** [comment]
The exception based GUI was a little unusual. It was used as I couldn't create a GUI in HOS.ApplicationLoader. 
Also, the emu restart was used as the method of disposing the emulation context and returning hanged the emu. (in testing)

**Comment by jms-c** [comment]
If anyone can share GUI code that doesn't use the weird exception approach, feel free to comment.

**Comment by jms-c** [comment]
If anyone wants to have a look at this code and change the ui. I will happily change it.
Have a nice day.

**Comment by jms-c** [comment]
Ping! If anyone wants to have a look, feel free!

**Comment by jms-c** [comment]
I have added the ability to stop the game without restarting the emulator!
Along with that, it now also doesn't use an exception to detect a dlc error.

Testers Needed!

**Comment by jms-c** [comment]
**Testing Information**
To test the latest commit, start a game with a moved dlc. 

When it prompts to continue, try to do one of the following:
Click "NO" and try to play another game
Click "YES"
Click "NO" and reopen the same game but click YES next
Or anything else regarding dlc.

Report any unusual errors (that don't happen normally)

**Comment by EmulationFanatic** [comment]
This is a great improvement! I tested the following scenarios:

With DLC intact:
Booted normally of course.

With DLC file moved:
- Launched and clicked NO - took me back to main window game selection screen as intended.
- Relaunched and clicked YES - launched the game without the missing DLC without issues.
- Relaunched and clicked NO - took me back to the main window where I then right-click Manage DLC. This warned me about missing DLC and advised to re-add it. Re-added it and launched successfully.

Awesome improvement!

**Comment by malfestio** [comment]
`00:00:38.857 |E| Application : DLC files have been moved or deleted; skipping 28 DLC files`
`00:00:38.857 |E| Application CreateErrorDialog: 28 DLC files have been moved or deleted; please re-add your DLC.`

works great, changed the folder name to mess with it an it gave me that error in a dialog window, changed folder back and it all appeared again

**Comment by jms-c** [comment]
On the discord, people have been having success with this pr.

**Comment by jms-c** [comment]
Waiting for reviews!
@AcK77 requested changes have been done!

**Comment by jms-c** [comment]
This PR now supports notifying when an update is moved/deleted!

**Comment by jms-c** [comment]
Please review!

**Review by AcK77 — CHANGES_REQUESTED** [review]
There is a lot of syntax issue, please fixes everything before asking for review...
You misses some spaces here and there like `if(blablabla>0)` should be `if (blablabla > 0)`, you added some empty lines where it's not needed and forget to add some where it's needed.

```
        public static int ignoreLoadErrorState = 0;
        public static int ignoreSelectedState = 0;
        public static string loadFailureString = null;
```
Those lines in ApplicationLoader is a bit strange. I don't think it's how things should be handled. But nice try, keep it up!

**Comment by jms-c** [comment]
It's unusual but needed to keep track of state.

**Comment by jms-c** [comment]
If not used, the pr is unable to differentiate whether to show DLC prompt, Update Prompt or none.

File: Ryujinx/Ui/Windows/TitleUpdateWindow.cs [hunk 5]
```diff
@@ -64,12 +65,23 @@ private TitleUpdateWindow(Builder builder, MainWindow parent, VirtualFileSystem
             }
 
             _baseTitleInfoLabel.Text = $"Updates Available for {titleName} [{titleId.ToUpper()}]";
-            
+
+            int invalidUpdateCount = 0;
             foreach (string path in _titleUpdateWindowData.Paths)
             {
+                if(!File.Exists(path))
```
Comment (riperiperi) [inline Ryujinx/Ui/Windows/TitleUpdateWindow.cs:72]: Missing space.
`if (!File.Exists(path))`

File: Ryujinx/Ui/Windows/TitleUpdateWindow.cs [hunk 6]
```diff
@@ -64,12 +65,23 @@ private TitleUpdateWindow(Builder builder, MainWindow parent, VirtualFileSystem
...
+                {
+                    invalidUpdateCount++;
+                }
```
Comment (riperiperi) [inline Ryujinx/Ui/Windows/TitleUpdateWindow.cs:75]: Maybe a newline after this.

File: Ryujinx/Ui/Windows/TitleUpdateWindow.cs [hunk 7]
```diff
@@ -64,12 +65,23 @@ private TitleUpdateWindow(Builder builder, MainWindow parent, VirtualFileSystem
...
                 AddUpdate(path);
```
Comment (riperiperi) [inline Ryujinx/Ui/Windows/TitleUpdateWindow.cs:76]: Is it alright that this update is still added, even when it is missing?

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 8]
```diff
@@ -66,6 +66,13 @@ public class ApplicationLoader
 
         public string TitleIdText => TitleId.ToString("x16");
 
+        // Track the state (Used to determine whether to show DLC Prompt/Update Prompt)
+        public static int ignoreLoadErrorState = 0;
```
Comment (riperiperi) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:70]: Public variables should start with a capital letter.
`IgnoreLoadErrorState`

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 9]
```diff
@@ -66,6 +66,13 @@ public class ApplicationLoader
...
+        // Used as a value for the button.
```
Comment (riperiperi) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:71]: What button?

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 10]
```diff
@@ -311,7 +327,10 @@ private void LoadNca(Nca mainNca, Nca patchNca, Nca controlNca)
             IFileSystem codeFs      = null;
 
             (Nca updatePatchNca, Nca updateControlNca) = GetGameUpdateData(_fileSystem, mainNca.Header.TitleId.ToString("x16"), _device.UserChannelPersistence.Index, out _);
-
+            if(loadFailureString != null)
```
Comment (riperiperi) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:330]: Needs a space in `if (`

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 11]
```diff
@@ -331,9 +350,31 @@ private void LoadNca(Nca mainNca, Nca patchNca, Nca controlNca)
...
+                    if (!File.Exists(dlcContainer.Path))
+                    {
+                        if (ignoreLoadErrorState < 2)
+                        {
+                            invalidDlcCount++;
+                        }
+                    }
+                }
+
+                if(invalidDlcCount > 0)
```
Comment (riperiperi) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:365]: Needs a space in `if (`

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 12]
```diff
@@ -331,9 +350,31 @@ private void LoadNca(Nca mainNca, Nca patchNca, Nca controlNca)
...
+                    ignoreSelectedState++;
+                    return;
+                }
 
                 foreach (DlcContainer dlcContainer in dlcContainerList)
                 {
+                    if (!File.Exists(dlcContainer.Path))
+                    {
+                        continue;
+                    }
```
Comment (riperiperi) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:377]: Newline after this.

File: Ryujinx/Ui/Windows/DlcWindow.cs [hunk 13]
```diff
@@ -92,6 +100,12 @@ private DlcWindow(Builder builder, VirtualFileSystem virtualFileSystem, string t
                     }
                 }
             }
+
+            if(invalidDlcCount > 0)
```
Comment (riperiperi) [inline Ryujinx/Ui/Windows/DlcWindow.cs:104]: Needs a space in `if (`

**Review by riperiperi — CHANGES_REQUESTED** [review]
I feel like the way the ApplicationLoader is returning errors isn't entirely desirable. I'm not a fan of returning strings destined for the UI from core emulator libraries like HLE, Armeilleure etc, as they will be harder to localize.

I think the way this should be done, is that the `LoadXci`, `LoadNca` etc functions could return a status code to indicate success, or a reason that they failed. You could also define a custom exception type and throw that, which will allow you to pass back additional information like the missing update/dlc count. Either solution would be a lot cleaner than using public fields as a means of returning method results.

I've commented on code style regardless so that you're aware of those.

File: Ryujinx/Ui/Windows/TitleUpdateWindow.cs (hunk 5 above)
Comment (jms-c) [reply Ryujinx/Ui/Windows/TitleUpdateWindow.cs:72]: Lol, genuinely didn't catch that because ryujinx is more lenient with title updates.

**Comment by jms-c** [comment]
Thanks for the helpful feedback, I will develop some fixes tomorrow.

**Comment by jms-c** [comment]
Syntax has improved and Rip In Peri Peri's suggestion has been implemented!

**Comment by jms-c** [comment]
Review requested.

**Comment by jms-c** [comment]
Please review.

**Comment by jms-c** [comment]
Conflicts resolved (hopefully)!

**Comment by jms-c** [comment]
Nope, not resolved lol. Time to code a fix.

**Comment by jms-c** [comment]
Fixed! It was literally a one line fix lol

**Comment by jms-c** [comment]
Converted to draft because it was hanging (and it's unmergeable therefore)

**Comment by jms-c** [comment]
I have attempted to fix hanging on this PR. However, I believe there is a better way (that I don't know atm)

**Comment by Shahil-Ayato** [comment]
Superseded by #2867.
